edgePy/DGEList.py

```diff
@@ -272,37 +308,143 @@ def library_size(self) -> np.array:
+        counts = self.counts[gene_mask].copy()
+
+        counts = (counts.T / gene_len_ordered).T
+        counts = counts / (col_sum / 1e6)
+
+        if transform_to_log:
+            counts = self.log_transform(counts, prior_count)
+            current_log = True
+
+        return self.copy(counts=counts, current_log=current_log, genes=genes)
+
+    def get_gene_mask_and_lengths(self, gene_data):
 
         """
-        raise NotImplementedError
+        use gene_data to get the gene lenths and a gene mask for the tranformation.
+        Args:
+            gene_data: the object that holds gene data from ensembl
 
-        # TODO: Implement here
-        # self = self.cpm(log=log, prior_count=prior_count)
+        """
+        gene_len_ordered = []
+        gene_mask = []
+        gene_ensg = []
+        for gene in self.genes:
+            if gene.startswith("ENSG"):
+                gene_name = gene
+                gene_ensg.append(gene_name)
+                if gene_data.has_gene(gene_name):
+                    gene_mask.append(True)
+                    gene_len_ordered.append(
+                        gene_data.get_length_of_canonical_transcript(gene_name) / 1e3
+                    )
+                else:
+                    gene_mask.append(False)
+            else:
+                t_gene = gene_data.get_genes_from_symbol(gene)
+                if t_gene:
+                    if len(t_gene) > 1:
+                        gene_name = gene_data.pick_gene_id(t_gene)
+                    else:
+                        gene_name = t_gene[0]
+                    gene_ensg.append(gene_name)
+                    if gene_data.has_gene(gene_name):
+                        gene_mask.append(True)
+                        gene_len_ordered.append(
+                            gene_data.get_length_of_canonical_transcript(gene_name) / 1e3
+                        )
+                    else:
+                        gene_mask.append(False)
+                else:
+                    gene_mask.append(False)
+        return gene_len_ordered, gene_mask
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        transform_to_log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> "DGEList":
+        """Normalize the DGEList to transcripts per million.
+
+        Adapted from Wagner, et al. 'Measurement of mRNA abundance using RNA-seq data:
+        RPKM measure is inconsistent among samples.' doi:10.1007/s12064-012-0162-3
+
+        Read counts :math:`X_i` (for each gene :math:`i` with gene length :math:`\widetilde{l_j}` ) are normalized as follows:
+        .. math::
+
+           TPM_i = \frac{X_i}{\widetilde{l_i}}\cdot (\frac{1}{\sum_j \frac{X_j}{\widetilde{l_j}}}) * 10^6
```

Comment:
@clintval can you help here? This one doesn't pass tox. Other variations like single line math environment don't work either. Error:
/home/moritz/Projects/edgePy/edgePy/DGEList.py:docstring of edgePy.DGEList.DGEList.tpm:10:Block quote ends without a blank line; unexpected unindent.

Reply:
I've included @clintval 's fixes to this comment.   Thanks @clintval !!